Support crate_type file extensions more fully

acmcarther opened the pull request:
Removes the hard coded rust_lib output, and declares the output file within the rust_library_impl. This lets us vary the file extension based on the toolchain -- OSX produces `.dylib` where linux would produce `.so`, for example.

Original description follows:
---
Tentatively addresses #36.

WIP: Need tests and understand platform implications

@davidzchen: Do we need the additional complexity of a hidden build rule? This change seems to work on my platform to produce dylibs that can be loaded correctly.

I've no idea how to handle different platforms. I added it as an additional attr, but I think a `config_setting` probably makes more sense.

davidzchen commented on rust/rust.bzl:
```diff
@@ -167,6 +188,19 @@ def _get_features_flags(features):
   return features_flags
 
+def _get_crate_type_and_target_outputs(name, crate_type, platform):
+  extension = LIBRARY_AND_PLATFORM_TO_EXTENSION[crate_type or "rlib", platform or "linux"]
```
Break up this line. See: https://bazel.build/versions/master/docs/skylark/bzl-style.html

acmcarther replied:
This is done (and was done a while ago)

Hoping posting this comment will complete the review.

davidzchen commented on rust/rust.bzl:
```diff
@@ -629,15 +663,14 @@ _rust_toolchain_attrs = {
-    outputs = {
-        "rust_lib": "lib%{name}.rlib",
-    },
+    outputs = _get_crate_type_and_target_outputs,
```
This does not seem idiomatic Skylark. Also, it is unclear to me how this would actually work where we pass in a function for the `outputs` parameter of `rule`. Is this an obscure Bazel/Skylark feature? I do not see it documented as a way to specify outputs in both the [Skylark documentation](https://bazel.build/versions/master/docs/skylark/rules.html#output-files) and the [Skylark Cookbook](https://bazel.build/versions/master/docs/skylark/cookbook.html#outputs-default).

@laurentlb @dslomov Can you comment on this?

damienmg replied:
This is a problem on our side, that should be documented. I believe @fweikert is the one who implemented the callback for outputs.

fweikert replied:
I didn't implement the feature (it has been in Bazel since April 2015), but I have a pending change that improves the documentation for implicit outputs in general (including callbacks).

Currently, the documentation of the "outputs" parameter mentions callbacks:
https://bazel.build/versions/master/docs/skylark/lib/globals.html#rule

davidzchen replied:
Ah! Good to know. Thanks for the clarification. Looking forward to the documentation.

acmcarther replied:
I was curious where I even got this from.... I made this change a while ago for my own use but couldn't remember how I'd found this feature, let alone find it again in the documentation.

Thanks all!

davidzchen commented on rust/rust.bzl:
```diff
@@ -25,6 +25,27 @@ LIBRARY_CRATE_TYPES = [
     "staticlib",
 ]
 
```
Also, please rebase and resolve merge conflicts.

damienmg commented:
Removing myself as David can perfectly manage this review

davidzchen commented:
@acmcarther This change LGTM and will be good go after rebasing and the minor style comment above. Thanks for working on this!

acmcarther commented:
@davidzchen Do you have any advice on resolving the platform issue? I'm not confident "platform" should be part of the build rules, but I'm also not sure I want to even solve the problem in this PR.

We could kick the can down the road by assuming everyone is linux... That's *still* an improvement over the current state, which just expected an rlib output...

davidzchen commented:
@acmcarther The only way I am aware of to get the host platform as a string would be to use [`ctx.fragments.cpp.cpu`](https://bazel.build/versions/master/docs/skylark/lib/fragments.html), but that requires `ctx`, which is only available in the rule implementation function. In this case, we want that information in the `outputs` callback. I also tried setting the `default` for the `platform` attribute to a `select` based on the `//rust:darwin` and `//rust:k8` `config_setting`s:

```python
_rust_library_attrs = {
    ...
    "platform": attr.string(
        default = select({
            "//rust:darwin": "darwin",
            "//rust:k8": "linux",
        })),
}
```

However, this does not work because `default` only accepts a `string`.

@damienmg, @fweikert - Is it currently possible to get this information from a `outputs` callback? If not, what is the most idiomatic way to implement this behavior?

acmcarther commented:
- Rebased
- Addressed comments
- Added dylib building rule

@davidzchen Rebase killed the commit your comment was on. Can you ack the review?

I'm pretty unsatisfied with the current platform situation, but I think we can improve the situation later.

acmcarther commented:
@davidzchen This is ready, but I can't address your review comment because the commit is gone. PTAL

davidzchen commented:
Sorry for the late reply. Let's go ahead and get this committed. Can you add a TODO comment to address the platforms issue in the previous comments and rebase?

acmcarther commented:
Rebased. I still cant clear up the "requested changes" snippet. Looks to be stuck...

acmcarther commented:
OK:

This is broken in a non-trivial way. The (new?) build bot is also building for darwin. This PR adds a rust_library of type "dylib" that is producing a `.dylib` file on darwin. This PR was attempting to kick platform detection down the road by including a `platform` attr. That's not going to work in this situation.

@davidzchen: Do you have any suggestions on proceeding? I can make the required change, but it might take a while. I don't think theres a "short cut" here -- rustc will *always* produce platform dependent `outs` for dylibs, and thats a sore point for Bazel.

mfarrugi commented:
This looks like it's also a blocker for dynamically linking cc_libs, as https://github.com/bazelbuild/rules_rust/blob/master/rust/rust.bzl#L169 also takes the shortcut of assuming only static linking.

Is this stalled on anything in particular, @davidzchen?

acmcarther commented:
This needs a second look to either make it smarter (w.r.t. platform extensions), or a trick to get the tests to pass.

I'll take a look at updating this PR this weekend.

mfarrugi commented:
@acmcarther I started adding support for linking to native dylibs, but `bazel run` and `bazel test` don't appear to pass LD_LIBRARY_PATH to the binary. 

Did this change actually support running or testing crates downstream of dylibs? It would be straightforward to add a wrapper similar to the bench rule, but I don't want to conflict with this change or redo work.

mfarrugi commented:
@acmcarther What were the test failures or the fix for this? CI doesn't have logs for it.

acmcarther commented:
mfarrugi@ Expected output file not produced under OSX. rustc is producing xyz.dylib, where xyz.so is expected.

Need to figure out a trick, perhaps using the toolchain, to dictate the expected output file.

acmcarther commented:
There are three moving parts as I see it here:

1) The selected toolchain
2) The desired output target
3) The desired crate type

(1) dictates the defaults for all crate types. (2) selects and overrides the values decided by (1), while (3) narrows the output down to a single possible value. (2) and (3) are already knowable within the output filepath determining callback, but I can't find any mechanism to determine (1). I [dropped a line in bazel-discuss](https://groups.google.com/forum/#!topic/bazel-discuss/fQuhTifRG1g) to see what might be available.
___

A totally different approach could be evaluated as well. `Rustc` does support an `-o OUTPUT_NAME` directive, which should let us force the file to have a consistent extension. This isn't easy to use for two reasons though
1) Using `-o` requires us to specify the full file path, not just as directory as we're currently doing. 
2) _I believe_ rustc relies on the output file extension to know how to handle the dependency. 

If (2) is true, we'd need to retain metadata with the rust_library to know how to turn it back into something rustc knows how to link.

acmcarther commented:
Apparently (from the bazel-discuss post) you do not need to specify the outputs directly -- you can specify them in the impl with a call to "declare_outputs". This lets us avoid the callback shenanegans entirely and defer this bit of planning until the impl. By the time we enter the impl, we have all the attrs and the toolchain.

Will update with this fix some time today.

acmcarther commented:
Tests now pass! @mfarrugi: PTAL at your earliest convenience. Hopefully this works with https://github.com/bazelbuild/rules_rust/pull/61 with minimal changes.

acmcarther commented on README.md:
```diff
@@ -112,7 +112,24 @@ rust_library(name, srcs, crate_root, deps, data, crate_features, rustc_flags)
+          The type of crate to be produced during library compilation. This
+          list closely matches Cargo's own notion of crate-type, and the
+          available options are "lib", "rlib", "dylib", "cdylib", "staticlib",
+          and "proc-macro".
```
This isn't evaluated as the required crate dependencies are impractical to import without `cargo-raze`.

For the record, the required crates are
```
extern crate proc_macro;
extern crate syn;
extern crate quote;
```

mfarrugi commented on rust/rust.bzl:
```diff
@@ -205,6 +198,27 @@ def _find_crate_root_src(srcs, file_names=["lib.rs"]):
+    extension = toolchain.dylib_ext
+  elif crate_type == "staticlib":
+    extension = toolchain.staticlib_ext
+  elif crate_type in ("rlib", "lib", "proc-macro"):
```
proc-macro should be a dylib (though I don't know if rustc actually cares)

acmcarther replied:
Ah this is important. We're relying on rustc to "produce the right thing" without us telling it what to make. Indeed, it is producing a dylib in this case, so we need to expect a dylib. Not sure how I became under the impression that this was an rlib -- my local code pre-osx fix used dylib here.

Fixed.

acmcarther commented:
ok to test

acmcarther commented:
@buchgr No problem, thanks for helping resolve the issue. 

I suspected it was some form of webui after leafing through https://github.com/bazelbuild/continuous-integration and not seeing a secret code phrase.

buchgr commented:
@acmcarther all future pull requests from you should trigger CI without a verification step. I have added an ACL passed on github collaborators. external contributors will require a project collaborator to approve it.  please file an issue on our continous-integration repo if that's not the case
